### PR TITLE
Add attack surface discoverer

### PR DESCRIPTION
**Pull request by julianoborba**

### CODE REVIEW

File: nosepass/attack_surface_discoverer.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+from nosepass.custom_json_logger import getLogger
+from nosepass.http_requestor import HTTPSession
+from nosepass.ports_analyser import \
+    scan_ports_with_nmap, \
+    get_ports_only_from_nmap_scan, \
+    retrieve_open_ports
+from socket import gethostbyname
+
+
+class HostObtainer:
```
Comment (rasecoiac03) [inline nosepass/attack_surface_discoverer.py]: I think there's no need to be a class.

File: nosepass/attack_surface_discoverer.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def discover(domain):
+    response = HTTPSession().do_get(domain)
+
+    server = ''
+    if 'Server' in response.headers:
+        server = response.headers['Server']
+
+    url = ''
```
Comment (rasecoiac03) [inline nosepass/attack_surface_discoverer.py]: 💅 

```suggestion
    url = response.url if response.url else ''
```

File: nosepass/attack_surface_discoverer.py (hunk 1 above)
Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: But on the other hand I won't be able to mock the  method return y-y

File: nosepass/attack_surface_discoverer.py (hunk 2 above)
Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: :+1:

File: nosepass/attack_surface_discoverer.py (hunk 1 above)
Comment (rasecoiac03) [reply nosepass/attack_surface_discoverer.py]: I see. Can't you do this `@patch('nosepass.attack_surface_discoverer.get_host_by_name')`?

I'm just trying to think how let things simple.

File: nosepass/attack_surface_discoverer.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class HostObtainer:
+
+    def get_host_by_name(self, domain):
+        try:
+            return gethostbyname(domain)
+        except Exception as socket_error:
+            getLogger().error(
+                'an error occurred while trying to get host'
```
Comment (shikow) [inline nosepass/attack_surface_discoverer.py]: How about change to `f-string` or `format`?

File: nosepass/attack_surface_discoverer.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                ' by name from domain %s, cause %s',
+                domain,
+                socket_error,
+                exc_info=1
```
Comment (shikow) [inline nosepass/attack_surface_discoverer.py]: Magic number :doge:

File: nosepass/attack_surface_discoverer.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            )
+            return ''
+
+
+def discover(domain):
+    response = HTTPSession().do_get(domain)
+
+    server = ''
+    if 'Server' in response.headers:
```
Comment (shikow) [inline nosepass/attack_surface_discoverer.py]: You can change to `response.headers.get('Server', '')`and avoid this if :)

File: nosepass/attack_surface_discoverer.py (hunk 2 above)
Comment (resilva87) [reply nosepass/attack_surface_discoverer.py]: @julianoborba same with `server =  response.headers['Server'] if 'Server' in response.headers else ''

File: nosepass/attack_surface_discoverer.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    host = HostObtainer().get_host_by_name(domain)
+
+    open_ports = []
+    if host:
+        nmap_result = scan_ports_with_nmap(host)
+        ports = get_ports_only_from_nmap_scan(nmap_result, host)
+        open_ports = retrieve_open_ports(ports)
+
+    return dict(
```
Comment (shikow) [inline nosepass/attack_surface_discoverer.py]: Guys, as we are using `Python 3.7` how about avoid create dicts and use `dataclass` with `type hint`?
ref. https://docs.python.org/3/library/dataclasses.html

File: nosepass/http_requestor.py [hunk 7]
```diff
@@ -2,43 +2,36 @@
...
-        getLogger().error(
-            'ConnectionError for %s, cause %s',
-            url,
-            connection_error,
-            exc_info=1
-        )
-        return Response()
+        except ConnectionError as connection_error:
+            getLogger().error(
+                'ConnectionError for %s, cause %s',
```
Comment (shikow) [inline nosepass/http_requestor.py]: How about change to `f-string` or `format`?

File: nosepass/attack_surface_discoverer.py (hunk 1 above)
Comment (shikow) [reply nosepass/attack_surface_discoverer.py]: 🤔

Comment (shikow) [reply nosepass/attack_surface_discoverer.py]: @julianoborba have you ever try to patch `socket.gethostbyname`?

Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: Yes, with no success... I really don't know why exactly, if isn't about being a static function... I will double check and call you just to ensure.

File: nosepass/attack_surface_discoverer.py (hunk 2 above)
Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: nice :+1:

File: nosepass/attack_surface_discoverer.py (hunk 3 above)
Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: :+1:

File: nosepass/attack_surface_discoverer.py (hunk 4 above)
Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: I will replace it

File: nosepass/attack_surface_discoverer.py (hunk 5 above)
Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: :+1:

File: nosepass/http_requestor.py (hunk 7 above)
Comment (julianoborba) [reply nosepass/http_requestor.py]: :+1:

File: nosepass/attack_surface_discoverer.py (hunk 6 above)
Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: Java feelings, I will use it

File: nosepass/attack_surface_discoverer.py (hunk 1 above)
Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: given the static function of the module, and not the class

```

def get_host_by_name(domain):
    try:
        return gethostbyname(domain)
    except Exception as socket_error:
        getLogger().error(
            f'an error occurred while trying to get host'
            f' by name from domain {domain}, cause {socket_error}',
            exc_info=IS_ENABLED
        )
        return ''

```
The `@patch('nosepass.attack_surface_discoverer.get_host_by_name')` works

But `@patch('socket.gethostbyname)` doesn't work.

It's ok to "wrap" the `gethostbyname` with a function defined by me, as above ?

**Comment by julianoborba** [comment]
Hey guys, please can you check again?

File: nosepass/attack_surface_discoverer.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+@dataclass
+class HostAttackSurface:
+    http_response = Response,
+    server_header = str,
+    response_url_location = str,
+    host = str,
+    open_ports = list
+
+    def __init__(self,
```
Comment (shikow) [inline nosepass/attack_surface_discoverer.py]: You dont need to specify `__init__` when you are using `dataclass` :) 
ref. https://docs.python.org/3/library/dataclasses.html#module-dataclasses

Comment (julianoborba) [reply nosepass/attack_surface_discoverer.py]: Got it

**Pull request merged**